File: Algorithms/130.SurroundedRegions.cpp
class Solution {
public:
    // 广搜。从上下左右四个边界往里走,凡是能碰到的'O',都是跟边界接壤的,应该保留
    void solve(vector<vector<char>>& board) {
        int m = board.size();
        if (m == 0)
            return;
        int n = board[0].size();
        for (int i=0; i < m; ++i) {
            bfs(board, i, 0);
            bfs(board, i, n-1);
        }
        for (int j=1; j < n-1; ++j) {
            bfs(board, 0, j);
            bfs(board, m-1, j);
        }
        for (int i=0; i < m; ++i) {
            for (int j=0; j < n; ++j) {
                if (board[i][j] == 'O')
                    board[i][j] = 'X';
                else if (board[i][j] == '+')
                    board[i][j] = 'O';
            }
        }
    }
    void bfs(vector<vector<char>> &board, int i, int j) {
        int n = board[0].size();
        queue<int> q;
        visit(board, i, j, q);
        while (!q.empty()) {
            int x = q.front(); q.pop();
            int ii = x / n;
            int jj = x % n;
            visit(board, ii-1, jj, q);
            visit(board, ii+1, jj, q);
            visit(board, ii, jj-1, q);
            visit(board, ii, jj+1, q);
        }
    }
    void visit(vector<vector<char>> &board, int i, int j, queue<int> &q) {
        int m = board.size();
        int n = board[0].size();
        if (i < 0 || j < 0 || i >= m || j >= n || board[i][j] != 'O')
            return;
        board[i][j] = '+';
        q.push(i * n + j);
    }
};

File: Algorithms/py/130.SurroundedRegions.py
class Solution(object):
    def solve(self, board):
        """
        :type board: List[List[str]]
        :rtype: void Do not return anything, modify board in-place instead.
        """
        m = len(board)
        if m == 0:
            return
        n = len(board[0])
        for i in range(m):
            self.bfs(board, i, 0)
            self.bfs(board, i, n-1)
        for j in range(1, n-1):
            self.bfs(board, 0, j)
            self.bfs(board, m-1, j)
        for i in range(m):
            for j in range(n):
                if board[i][j] == 'O':
                    board[i][j] = 'X'
                elif board[i][j] == '+':
                    board[i][j] = 'O'

    def bfs(self, board, i, j):
        n = len(board[0])
        q = []
        self.visit(board, i, j, q)
        while q:
            x = q.pop(0)
            ii = x / n
            jj = x % n
            self.visit(board, ii-1, jj, q)
            self.visit(board, ii+1, jj, q)
            self.visit(board, ii, jj-1, q)
            self.visit(board, ii, jj+1, q)

    def visit(self, board, i, j, q):
        m = len(board)
        n = len(board[0])
        if i < 0 or j < 0 or i >= m or j >= n or board[i][j] != 'O':
            return
        board[i][j] = '+';
        q.append(i * n + j)
